src/main/java/com/yahoo/bullet/bql/query/QueryBuilder.java

```diff
@@ -0,0 +1,612 @@
+        if (whereNode != null) {
+            filter = visit(processedQuery.getWhere());
+            if (cannotCastToBoolean(filter.getType())) {
+                addError(whereNode, "WHERE clause cannot be casted to BOOLEAN: " + whereNode, "Please specify a valid WHERE clause.");
+            }
+        }
+    }
+
+    public void doSelect() {
+        doSelectFields();
+
+        requiresNoCopyFlag = true;
+
+        addSchemaLayer(true);
+
+        aggregation = new Raw(limit);
+
+        // Missing fields
+        Set<String> additionalFields = OrderByProcessor.visit(processedQuery.getOrderByNodes(), layeredSchema, baseSchema);
+
+        for (String field : additionalFields) {
+            FieldExpression expression = new FieldExpression(field);
+            expression.setType(layeredSchema.getType(field));
+            addProjectionField(field, expression);
+        }
+
+        doOrderBy();
+
+        if (!additionalFields.isEmpty()) {
+            postAggregations.add(new Culling(additionalFields));
+        }
+
+        // Create projection at the end because ORDER BY can add additional fields
+        doProjection();
+    }
+
+    public void doSelectAll() {
+        doSelectFields();
+
+        requiresCopyFlag = processedQuery.getSelectNodes().stream().anyMatch(node -> !(node instanceof FieldExpressionNode) || processedQuery.hasAlias(node));
+
+        doProjection();
+
+        addSchemaLayer(requiresCopyFlag);
+
+        aggregation = new Raw(limit);
+
+        doOrderBy();
+
+        // Renamed fields that are not in the final schema are removed
+        if (requiresCopyFlag) {
+            Schema schema = layeredSchema.getSchema();
+            Set<String> transientFields = layeredSchema.getAliases().keySet().stream()
+                                                                             .filter(field -> !schema.hasField(field))
+                                                                             .collect(Collectors.toCollection(HashSet::new));
+            if (!transientFields.isEmpty()) {
+                postAggregations.add(new Culling(transientFields));
+            }
+        }
+    }
+
+    private void doSelectDistinct() {
+        Map<String, String> fields = new HashMap<>();
+        for (ExpressionNode node : processedQuery.getSelectNodes()) {
+            Expression expression = visit(node);
+            String newName = processedQuery.getAliasOrName(node);
+            String name = node.getName();
+            Type type = expression.getType();
+            if (processedQuery.hasAlias(node)) {
+                addAlias(name, newName);
+            }
+            addProjectionField(name, expression);
+            addSchemaField(newName, type);
+
+            fields.put(name, newName);
+
+            if (isNonPrimitive(type)) {
+                addError(node, "The SELECT DISTINCT field " + node + " is non-primitive. Type given: " + type, "Please specify primitive fields only for SELECT DISTINCT.");
```

Comment:
Can we enum these errors and have some toString like thing in the enum that takes a list of Objects to create this message? Can table for later if you don't want to do that now